generators/modelIntensity.cc

```diff
@@ -0,0 +1,210 @@
+	const int         waveIndex = _fitResult->waveIndex(waveName);
+	if (waveIndex == -1) {
+		printErr << "wave '" << waveName << "' not used in fit result provided. Aborting..." << std::endl;
+		return false;
+	}
+	if (_amplitudes[waveIndex]) {
+		printErr << "amplitude for wave '" << waveName << "' is added a second time. Aborting..." << std::endl;
+		return false;
+	}
+
+	_amplitudesInitialized = false;
+	_amplitudes[waveIndex] = amplitude;
+	_refls     [waveIndex] = amplitude->decayTopology()->XIsobarDecayVertex()->parent()->reflectivity();
+	_allRefls.insert(amplitude->decayTopology()->XIsobarDecayVertex()->parent()->reflectivity());
+	return true;
+}
+
+
+bool
+rpwa::modelIntensity::addIntegral(const rpwa::ampIntegralMatrix& integralMatrix)
+{
+	_integralsLoaded = false;
+	_integrals.resize(_fitResult->nmbWaves());
+
+	for (size_t wave = 0; wave < _fitResult->nmbWaves(); ++wave) {
+		const std::string& waveName = _fitResult->waveName(wave);
+		if (waveName == "flat") {
+			_integrals[wave] = 1.;
+		} else {
+			if (not integralMatrix.containsWave(waveName)) {
+				std::cout << "wave '" << waveName << "' not in the integral matrix. Aborting..." << std::endl;
+				return false;
+			}
+			const std::complex<double> integral = integralMatrix.element(waveName, waveName);
+			_integrals[wave] = std::sqrt(integral.real());
+		}
+	}
+
+	_integralsLoaded = true;
+	return true;
+}
+
+
+bool
+rpwa::modelIntensity::initAmplitudes(const std::vector<std::string>& decayKinParticleNames)
+{
+	// get X name from first amplitude
+	std::vector<std::string> prodKinParticleNames(1);
+	for (size_t wave = 0; wave < _fitResult->nmbWaves(); ++wave) {
+		const std::string& waveName = _fitResult->waveName(wave);
+		// no amplitude for 'flat' wave
+		if (waveName == "flat") {
+			if (_amplitudes[wave]) {
+				printErr << "amplitude for flat wave was added. Aborting..." << std::endl;
+				return false;
+			}
+			continue;
+		}
+		if (not _amplitudes[wave]) {
+			printErr << "amplitude for wave '" << waveName << "' was not added. Aborting..." << std::endl;
+			return false;
+		}
+
+		const isobarDecayTopologyPtr decay  = _amplitudes[wave]->decayTopology();
+		const isobarDecayVertexPtr   vertex = decay->XIsobarDecayVertex();
+		prodKinParticleNames[0] = vertex->parent()->name();
+		break;
+	}
+
+	return initAmplitudes(prodKinParticleNames, decayKinParticleNames, true);
+}
+
+
+bool
+rpwa::modelIntensity::initAmplitudes(const std::vector<std::string>& prodKinParticleNames,
+                                     const std::vector<std::string>& decayKinParticleNames,
+                                     const bool                      fromXDecay)
+{
+	_amplitudesInitialized = false;
+	_amplitudesFromXDecay  = fromXDecay;
+
+	for (size_t wave = 0; wave < _fitResult->nmbWaves(); ++wave) {
+		const std::string& waveName = _fitResult->waveName(wave);
+		// no amplitude for 'flat' wave
+		if (waveName == "flat") {
```

Comment:
checks duplicated in `rpwa::modelIntensity::initAmplitudes(const std::vector<std::string>& decayKinParticleNames)`
are they needed there?